Shopify/ruby-lsp
# Proposal for expectations testing

### Motivation

Simplify the way we write tests.

### How to add a new test for an existing request handler

(edited for the last iteration)

1. Add a new fixture `my_fixture.rb` file under `test/fixtures`
2. (optional) Add new expectations under `test/expectations/$HANDLER` for the request handlers you're concerned by
3. Profit by running `bin/test test/requests/$HANDLER_expectations_test my_fixture`
    * Handlers for which you added expectations will be checked with `assert_expectations`
    * Handlers without expectations will be ran against your new test to check that nothing breaks
    
### How to add a new request handler

1. Add a new file under `test/requests` that subclasses `ExpectationsTestRunner` and call `expectations_tests $HANDLER, "$EXPECTATION_DIR"` where: `$HANDLER` is the fully qualified name or your handler class and `$EXPECTATION_DIR` is the directory name where you want to store the expectation files.

2. (optional) Override the `run_expectations` and `assert_expectations` methods if needed. See the different request handler expectation runner for examples.

4. (optional) Add new fixtures for your handler under `test/fixtures`

5. (optional) Add new expectations under `test/expectations/$HANDLER`
   * No need to write the expectations by hand, just run the test with an empty expectation file and copy from the output.

7. Profit by running, `bin/test test/expectations_test $HANDLER`
    * Tests with expectations will be checked with `assert_expectations`
    * Tests without expectations will be ran against your new $HANDLER to check that nothing breaks

### TODO

* [ ] I think we can simplify the test output a bit
* [ ] We may be able to simplify how we write the `run_x`, `assert_x` to not have to duplicate code
* [ ] We need to somehow catch the stderr of `Formatting` to avoid cluttering the output
* [ ] Migrate `SemanticHighlighting` but I wanted to show that just adding the mechanics already caught two errors:

   <img width="933" alt="image" src="https://user-images.githubusercontent.com/583144/162322086-253b7c41-ac88-4d52-9ccb-b6d5b8c7f269.png">
